src/cache.c

```diff
@@ -67,26 +72,57 @@ extern Cache *
 cache_pin(Cache *cache)
 {
 	MemoryContext old = MemoryContextSwitchTo(CacheMemoryContext);
+	CachePin   *cp = palloc(sizeof(CachePin));
 
-	pinned_caches = lappend(pinned_caches, cache);
+	cp->cache = cache;
+	cp->subtxnid = GetCurrentSubTransactionId();
+	pinned_caches = lappend(pinned_caches, cp);
 	MemoryContextSwitchTo(old);
 	cache->refcount++;
 	return cache;
 }
 
-extern int
-cache_release(Cache *cache)
+static void
+remove_pin(Cache *cache, SubTransactionId subtxnid)
+{
+	ListCell   *lc,
+			   *prev = NULL;
+
+	foreach(lc, pinned_caches)
+	{
+		CachePin   *cp = lfirst(lc);
+
+		if (cp->cache == cache && cp->subtxnid == subtxnid)
+		{
+			pinned_caches = list_delete_cell(pinned_caches, lc, prev);
+			return;
+		}
+
+		prev = lc;
+	}
+
+	Assert(false);
```

Comment:
Maybe add comment that we shouldn't reach here because a pin should always be removed?